src/compiler/crystal/codegen/primitives.cr

```diff
@@ -326,6 +326,21 @@ class Crystal::CodeGenVisitor
     )
   end
 
+  private def codegen_out_of_range(target_type : IntegerType, arg_type : FloatType, arg)
+    if arg_type.kind == :f32 && target_type.kind == :u128
+      # since Float32::MAX < UInt128::MAX
+      # the range checking is replaced by a positive check only
+      builder.fcmp(LLVM::RealPredicate::OLT, arg, llvm_type(arg_type).const_float(0))
```

Comment:
Can't we generalize this to `arg_type.max < target_type.max`? I'd prefer express it in terms of that than explicitly code the types.

Reply:
That would require the compiler to support comparing 128 bits with float or doubles. And the comparison of u128 and f32 will convert the u128 to f32 so we first will need to force everything to be f64. (Actually I think that `Int128#<(Float32)` might not be sound)

On other side these check are in order to emit as few instructions as possible in the runtime check.

And there is a need for specific instructions on each types combinations.

I feel the pain :-) but I think it needs to remain as is.

Reply:
I'm not sure I understand.

Reply:
I don't see that checking `arg_type.max < target_type.max` is a generalization. Because, without checking the inner code of the `then` side, that assume that all the cases are of non negative integer types.

Making all (or this) `codegen_out_of_range` more general would require to check the number types ranges probably in double. I prefer to keep the code using only the type involving in the check (even if it's a bit less declarative). I think it's easier to follow.

Reply:
OK, I can imagine how this method would get more complex if you had to check `target_type.min < arg_type.min` too.

Reply:
For the record: I feel the pain :-)